File: src/pcl-bindings/StatisticalOutlierRemovalPointXYZ.cpp
#include <pcl/filters/statistical_outlier_removal.h>
#include "pcl/impl/point_types.hpp"
#include "pcl/point_cloud.h"
#include "export.h"

using namespace pcl;
using point_t = PointXYZ;
using pointcloud_t = PointCloud<point_t>;

EXPORT(StatisticalOutlierRemoval<point_t>*) statistical_outlier_removal_pointxyz_ctor()
{
    return new StatisticalOutlierRemoval<point_t>();
}

EXPORT(void) statistical_outlier_removal_pointxyz_delete(StatisticalOutlierRemoval<point_t>** ptr)
{
    delete* ptr;
    *ptr = NULL;
}

EXPORT(void) statistical_outlier_removal_pointxyz_set_input_cloud(StatisticalOutlierRemoval<point_t>* ptr, pointcloud_t* cloud)
{
    pointcloud_t::Ptr shared(std::make_shared<pointcloud_t>(*cloud));
    ptr->setInputCloud(shared);
}

EXPORT(void) statistical_outlier_removal_pointxyz_set_mean_k(StatisticalOutlierRemoval<point_t>* ptr, int mean_k)
{
    ptr->setMeanK(mean_k);
}

EXPORT(int) statistical_outlier_removal_pointxyz_get_mean_k(StatisticalOutlierRemoval<point_t>* ptr)
{
    return ptr->getMeanK();
}

EXPORT(void) statistical_outlier_removal_pointxyz_set_stddev_mul_thresh(StatisticalOutlierRemoval<point_t>* ptr, double stddev_mul_thresh)
{
    ptr->setStddevMulThresh(stddev_mul_thresh);
}

EXPORT(double) statistical_outlier_removal_pointxyz_get_stddev_mul_thresh(StatisticalOutlierRemoval<point_t>* ptr)
{
    return ptr->getStddevMulThresh();
}

EXPORT(void) statistical_outlier_removal_pointxyz_filter(StatisticalOutlierRemoval<point_t>* ptr, pointcloud_t* output)
{
    ptr->filter(*output);
}

EXPORT(void) statistical_outlier_removal_pointxyz_set_filter_indices_vector(StatisticalOutlierRemoval<point_t>* ptr, std::vector<int>* indices)
{
    pcl::PointIndices::Ptr indices_ptr(new pcl::PointIndices());
    indices_ptr->indices = *indices;
    ptr->setIndices(indices_ptr);
}

EXPORT(void) statistical_outlier_removal_pointxyz_get_filter_indices_vector(StatisticalOutlierRemoval<point_t>* ptr, std::vector<int>* indices)
{
    pcl::IndicesPtr indices_ptr = ptr->getIndices();
    indices->assign(indices_ptr->begin(), indices_ptr->end());
}

EXPORT(void) statistical_outlier_removal_pointxyz_set_indices(StatisticalOutlierRemoval<point_t>* ptr, std::size_t row_start, std::size_t col_start, std::size_t nb_rows, std::size_t nb_cols)
{
    ptr->setIndices(row_start, col_start, nb_rows, nb_cols);
}

File: src/Pcl.NET/StatisticalOutlierRemovalPointXYZ.cs
namespace Pcl.NET
{
    /// <summary>
    /// Removes points whose mean distance to their k nearest neighbors is beyond a standard deviation threshold.
    /// General-purpose noise removal, lidar scans, structured point clouds.
    /// </summary>
    public class StatisticalOutlierRemovalPointXYZ : Filter<PointXYZ>
    {
        private PointCloud<PointXYZ>? _input;
        private VectorInt _indices = new VectorInt();

        public StatisticalOutlierRemovalPointXYZ()
        {
            _ptr = Invoke.statistical_outlier_removal_pointxyz_ctor();
        }

        public override PointCloud<PointXYZ>? Input
        {
            get => _input;
            set
            {
                _input = value;
                Invoke.statistical_outlier_removal_pointxyz_set_input_cloud(_ptr, value);
            }
        }

        public override VectorInt Indices
        {
            get
            {
                ThrowIfDisposed();
                Invoke.statistical_outlier_removal_pointxyz_get_filter_indices_vector(_ptr, _indices);
                return _indices;
            }
            set
            {
                ThrowIfDisposed();
                _indices = value;
                Invoke.statistical_outlier_removal_pointxyz_set_filter_indices_vector(_ptr, value);
            }
        }

        /// <summary>
        /// Get or Set the number of nearest neighbors to use for mean distance estimation. Higher K → more stable, slower
        /// </summary>
        public int MeanK
        {
            get
            {
                ThrowIfDisposed();
                return Invoke.statistical_outlier_removal_pointxyz_get_mean_k(_ptr);
            }
            set
            {
                ThrowIfDisposed();
                Invoke.statistical_outlier_removal_pointxyz_set_mean_k(_ptr, value);
            }
        }
        /// <summary>
        /// Get or Set the standard deviation multiplier for the distance threshold calculation.
        /// The distance threshold will be equal to: mean + stddev_mult* stddev.
        /// Points will be classified as inlier or outlier if their average neighbor distance is below or above this threshold respectively.
        /// Lower → more aggressive removal
        /// </summary>
        public float StddevMulThresh
        {
            get
            {
                ThrowIfDisposed();
                return (float)Invoke.statistical_outlier_removal_pointxyz_get_stddev_mul_thresh(_ptr);
            }
            set
            {
                ThrowIfDisposed();
                Invoke.statistical_outlier_removal_pointxyz_set_stddev_mul_thresh(_ptr, value);
            }
        }

        public override void SetIndices(long row_start, long col_start, long nb_rows, long nb_cols)
        {
            ThrowIfDisposed();

            ThrowIfBadIndices(row_start, col_start, nb_rows, nb_cols);

            Invoke.statistical_outlier_removal_pointxyz_set_filter_indices(_ptr, (ulong)row_start, (ulong)col_start, (ulong)nb_rows, (ulong)nb_cols);
        }

        public override PointCloud<PointXYZ> ApplyFilter()
        {
            ThrowIfDisposed();
            ThrowIfInputNotSet();
            var output = new PointCloudXYZ();
            Invoke.statistical_outlier_removal_pointxyz_filter(_ptr, output);
            return output;
        }

        protected override void DisposeObject()
        {
            if (!_suppressDispose)
            {
                Invoke.statistical_outlier_removal_pointxyz_delete(ref _ptr);
                _ptr = IntPtr.Zero;
            }
        }
    }
}
